numpy/core/arrayprint.py

```diff
@@ -1164,6 +1184,11 @@ def array_str(a, max_line_width=None, precision=None, suppress_small=None):
     '[0 1 2]'
 
     """
+    if a.shape == () and not a.dtype.names:
+        if _format_options['legacy']:
+            return str(a.item())
+        return str(a[()])
```

Comment:
What effect is guarding this under `a.dtype.names` in the non-legacy case having?

Reply:
I'd perhaps be inclined to write this as the slightly longer

```
if _format_options['legacy'] and a.shape == () and not a.dtype.names:
    return str(a.item())

# actually justify this behavior in a comment here, where its a feature, not deprecated behaviour
if a.shape == () and not a.dtype.names:
    return str(a[()])
```
just to separate the legacy conditions from the rest

Reply:
Yeah good idea for clarity.

I had the same thought about `a.dtype.names` when writing this. I did confirm for myself that in 1.13 the `a.dtype.names` case was special-cased to use the `formatter` to represent the value, i.e. it should go through the `array2string` code, which uses `formatter`.

Reply:
Hmm but actualy, here is the implication of excluding `a.dtype.names` in the non-legacy case.

If we do `if a.shape == () and not a.dtype.names:`, then the eventual output is the output of `StructureFormat`, but with the `precision` specified as argument to `array_str`, but all other settings from `__format_options`. I am not sure the `precision` argument is ever used.

If we only do `if a.shape == ()`, then the eventual output is still the output of `StructureFormat`, but with  all settings, including `precision`, from `__format_options`. We go through a different path through `void_scalar_repr` to get there but the result is almost identical.

I'm going to remove the `a.dtype.names` test in the non-legacy case, just to make the code simpler. I don't think it has any real effect. I'll leave it in legacy mode since  it makes a difference there,